{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dependent-person",
   "metadata": {},
   "source": [
    "# Branch and Bound\n",
    "## Specification\n",
    "- Name: Template for Problem 16, 17, 18 in Rosalind\n",
    "- Name your notebooks as: problem16.ipynb, problem17.ipynb, problem18.ipynb\n",
    "- options: none\n",
    "- input: filename passed as first parameter to main\n",
    "- output: a text file. ( using print ( .... file=someFileObject) is a handy way to do this after you have opened someFileObject as a text file). I find it handy to name these files by creating a string by concatenating the string named infile with \".out\" ... rosalind4.txt.out ( for example).\n",
    "-Rosalind Problem Names:\n",
    "    - Find Substrings of a Genome Encoding a Given Amino Acid String\n",
    "    - Generate the Theoretical Spectrum of a Cyclic Peptide\n",
    "    - Find a Cyclic Peptide with Theoretical Spectrum Matching an Ideal Spectrum\n",
    "\n",
    "As always, include an Inspection Intro Markdown that describes your specific algorithm at the beginning of the notebook, and another Inspection Results markdown at the end of the notebook that documents: your inspection team, the findings of the team, and your resolution of those findings.\n",
    "\n",
    "Please submit your three notebooks, an example of one of the Rosalind files that you ran and passed, and the output that your program generated as a text file.\n",
    "\n",
    "## Description\n",
    "These are drawn from material presented in Ch. 4 of Compeau and Pevzner, with a focus on the Branch and Bound algorithm.\n",
    "\n",
    "## Hints\n",
    "1) Make sure to write your code with a well defined class that you instantiate an object from in your main() function\n",
    "Here is a template to consider."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "portuguese-management",
   "metadata": {},
   "source": [
    "## Inspection Intro"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d26694da",
   "metadata": {},
   "source": [
    "To solve this problem we:\n",
    "1. Determine the length of the DNA sequence of the target peptide by multiplying it by 3, because each amino acid is 3 bases. Call this length N\n",
    "2. We iterate the input DNA sequence, scanning each reading frame generating chunks of size N. We translate the chunk to its peptide chain and check if it matches our target.\n",
    "3. Save the matches and then output findings after iterating the DNA sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "photographic-vietnamese",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGTCGGTAGGCTGCCCCATGATTG\n",
      "CAATCATGGGGATCTTTGCCTACG\n",
      "AGTCGGAAGTGAACCCCATGATTG\n",
      "CAAAGTTGGGGCTCGTTACCGACG\n",
      "GGTTGGAAGAGAACCCCAAGATTG\n",
      "CAGTCGTGGGGTTCACTACCAACA\n",
      "CGTTGGCAGGGAGCCCCAACTCTG\n",
      "AGTAGGCAGCGAACCCCATGACTG\n",
      "CAGTCCTGGGGATCTCTACCCACT\n",
      "CAGTCGTGGGGGTCCTTGCCAACG\n",
      "GGTAGGTAGAGAGCCCCATGATTG\n",
      "CAGAGTTGGGGAAGTCTACCAACG\n",
      "TGTGGGCAGGGAACCCCAGCTCTG\n",
      "GGTAGGAAGACTTCCCCAGGATTG\n",
      "CGTCGGCAGCGATCCCCAGCTCTG\n",
      "TGTGGGTAAGGATCCCCAACTTTG\n",
      "TGTGGGCAGAGATCCCCAACTTTG\n",
      "CAGTCATGGGGTTCACTCCCGACG\n",
      "CAATCCTGGGGCTCGCTCCCGACT\n",
      "GGTTGGTAGCGAACCCCAGCTTTG\n"
     ]
    }
   ],
   "source": [
    "def reverseComplement(seq: str) -> str:\n",
    "        \"\"\"Generates reverse complement for given sequence.\n",
    "\n",
    "        Args:\n",
    "            seq (str): Kmer of alphabet {A,G,C,T}\n",
    "\n",
    "        Returns:\n",
    "            str: Reverse complement of input seq\n",
    "        \"\"\"\n",
    "        return seq.translate(str.maketrans(\"AGCT\", \"TCGA\"))[::-1]\n",
    "\n",
    "class Peptides:\n",
    "    \"\"\"\n",
    "    Class Peptides for solving Rosalind #16\n",
    "\n",
    "    Parameters: \n",
    "        seq: str - The DNA string\n",
    "    \"\"\"\n",
    "    def __init__(self, seq):\n",
    "        self.seq = seq\n",
    "        self.revSeq = reverseComplement(seq)\n",
    "        self.rnaCodonTable = {\n",
    "        # RNA codon table\n",
    "        # U\n",
    "        'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
    "        'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
    "        'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
    "        'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
    "        # C\n",
    "        'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
    "        'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
    "        'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
    "        'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
    "        # A\n",
    "        'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
    "        'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
    "        'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
    "        'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
    "        # G\n",
    "        'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
    "        'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
    "        'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
    "        'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'  # GxG\n",
    "    }\n",
    "        self.dnaCodonTable = {key.replace('U', 'T'): value for key, value in self.rnaCodonTable.items()}\n",
    "    def translateDnaToAA(self, dna):\n",
    "        \"\"\"\n",
    "        Transaltes DNA to codons then finally to amino acids.\n",
    "        \"\"\"\n",
    "        aa = \"\"\n",
    "        x = 0 \n",
    "        while x < len(dna):\n",
    "            s = dna[x:x+3]\n",
    "            a = self.dnaCodonTable[s]\n",
    "            aa += a \n",
    "            x += 3\n",
    "        return aa\n",
    "    def findOccurences(self, peptide):\n",
    "        \"\"\"\n",
    "        Finds occurences of input peptide in input sequence.\n",
    "        \"\"\"\n",
    "        found = []\n",
    "        \n",
    "        peptideSeqLen = len(peptide)*3\n",
    "        # iterate each reading frame\n",
    "        for frame in range(0,3):\n",
    "            p = frame\n",
    "            while p <= len(self.seq):\n",
    "                # generate candidate seq\n",
    "                candidateSeq = self.seq[p: p+peptideSeqLen]\n",
    "                if len(candidateSeq) == peptideSeqLen:\n",
    "                    # translate to peptide\n",
    "                    revCandidateSeq = reverseComplement(candidateSeq)\n",
    "                    forwardPep = self.translateDnaToAA(candidateSeq)\n",
    "                    reversePep = self.translateDnaToAA(revCandidateSeq)\n",
    "                    # check peptides against target\n",
    "                    if forwardPep == peptide:\n",
    "                        found.append(candidateSeq)\n",
    "\n",
    "                    if reversePep == peptide:\n",
    "                        found.append(candidateSeq)\n",
    "                p += 3\n",
    "        \n",
    "        return found\n",
    "\n",
    "\n",
    "\n",
    "def main(inFile = None):\n",
    "    '''\n",
    "    Do the main thing\n",
    "    '''\n",
    "    with open(inFile) as fh:\n",
    "        lines = [l.strip() for l in fh.readlines()]\n",
    "        seq = lines[0]\n",
    "        targetPeptide = lines[1]\n",
    "    \n",
    "    peptide = Peptides(seq)\n",
    "    with open(\"cmirchandani_rosalind16_out.txt\", \"w\") as f:\n",
    "        for p in peptide.findOccurences(targetPeptide):\n",
    "            print(p)\n",
    "            print(p, file=f)\n",
    "    \n",
    "    \n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile = 'rosalind_ba4b.txt') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "paperback-warning",
   "metadata": {},
   "source": [
    "## Inspection Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3d4983d",
   "metadata": {},
   "source": [
    "Inspection group: Gabe P., Jodie J.,\n",
    "\n",
    "- Added print to stdout as well as file\n",
    "- Added more inline comments to explain code \n",
    "- Removed unused code"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13 (main, May 24 2022, 21:13:51) \n[Clang 13.1.6 (clang-1316.0.21.2)]"
  },
  "vscode": {
   "interpreter": {
    "hash": "b0fa6594d8f4cbf19f97940f81e996739fb7646882a419484c72d19e05852a7e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
