{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dependent-person",
   "metadata": {},
   "source": [
    "# Branch and Bound\n",
    "## Specification\n",
    "- Name: Template for Problem 16, 17, 18 in Rosalind\n",
    "- Name your notebooks as: problem16.ipynb, problem17.ipynb, problem18.ipynb\n",
    "- options: none\n",
    "- input: filename passed as first parameter to main\n",
    "- output: a text file. ( using print ( .... file=someFileObject) is a handy way to do this after you have opened someFileObject as a text file). I find it handy to name these files by creating a string by concatenating the string named infile with \".out\" ... rosalind4.txt.out ( for example).\n",
    "-Rosalind Problem Names:\n",
    "    - Find Substrings of a Genome Encoding a Given Amino Acid String\n",
    "    - Generate the Theoretical Spectrum of a Cyclic Peptide\n",
    "    - Find a Cyclic Peptide with Theoretical Spectrum Matching an Ideal Spectrum\n",
    "\n",
    "As always, include an Inspection Intro Markdown that describes your specific algorithm at the beginning of the notebook, and another Inspection Results markdown at the end of the notebook that documents: your inspection team, the findings of the team, and your resolution of those findings.\n",
    "\n",
    "Please submit your three notebooks, an example of one of the Rosalind files that you ran and passed, and the output that your program generated as a text file.\n",
    "\n",
    "## Description\n",
    "These are drawn from material presented in Ch. 4 of Compeau and Pevzner, with a focus on the Branch and Bound algorithm.\n",
    "\n",
    "## Hints\n",
    "1) Make sure to write your code with a well defined class that you instantiate an object from in your main() function\n",
    "Here is a template to consider."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "portuguese-management",
   "metadata": {},
   "source": [
    "## Inspection Intro"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e131ada8",
   "metadata": {},
   "source": [
    "To solve this problem:\n",
    "1. Generate a list of candidate peptides by keeping a list of list of peptide masses.\n",
    "2. Check if each candidate peptide matches the target spectrum, or if the candidate could be a possible match by checking if its a subset of the target spectrum.\n",
    "3. If it matches we keep it and remove from candidates. If it doesn't match and is not consistent, remove from candidates. If it is consistent we keep it in the cadidates list.\n",
    "4. Repeat until no more candidates exist. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "photographic-vietnamese",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "97-101-87-156-71-103-114-97-115 97-114-103-71-156-87-101-97-115 71-103-114-97-115-97-101-87-156 101-97-115-97-114-103-71-156-87 103-71-156-87-101-97-115-97-114 115-97-101-87-156-71-103-114-97 87-101-97-115-97-114-103-71-156 156-87-101-97-115-97-114-103-71 156-71-103-114-97-115-97-101-87 114-97-115-97-101-87-156-71-103 114-103-71-156-87-101-97-115-97 71-156-87-101-97-115-97-114-103 103-114-97-115-97-101-87-156-71 115-97-114-103-71-156-87-101-97 101-87-156-71-103-114-97-115-97 97-115-97-114-103-71-156-87-101 97-115-97-101-87-156-71-103-114 87-156-71-103-114-97-115-97-101\n"
     ]
    }
   ],
   "source": [
    "def reverseComplement(seq: str) -> str:\n",
    "        \"\"\"Generates reverse complement for given sequence.\n",
    "\n",
    "        Args:\n",
    "            seq (str): Kmer of alphabet {A,G,C,T}\n",
    "\n",
    "        Returns:\n",
    "            str: Reverse complement of input seq\n",
    "        \"\"\"\n",
    "        return seq.translate(str.maketrans(\"AGCT\", \"TCGA\"))[::-1]\n",
    "\n",
    "class Peptides:\n",
    "    \"\"\"\n",
    "    Class Peptides for solving Rosalind #18\n",
    "\n",
    "    Parameters: \n",
    "        spectrum: str - The input peptide spectrum\n",
    "    \"\"\"\n",
    "    def __init__(self, spectrum):\n",
    "        self.s = list(spectrum.split(' '))\n",
    "        self.spectrum = list(map(int, self.s))\n",
    "        self.massTable = {\n",
    "            'G': 57, 'A': 71, 'S': 87,\n",
    "            'P': 97, 'V': 99, 'T': 101,\n",
    "            'C': 103, 'I': 113, 'L': 113,\n",
    "            'N': 114, 'D': 115, 'K': 128,\n",
    "            'Q': 128, 'E': 129, 'M': 131,\n",
    "            'H': 137, 'F': 147, 'R': 156,\n",
    "            'Y': 163, 'W': 186\n",
    "        }\n",
    "    \n",
    "    def cylcoSpectrum(self, seq):\n",
    "        \"\"\"\n",
    "        Generates the cyclospectrum given a list of masses of a cyclic peptide.\n",
    "        Parameters:\n",
    "            seq: list[int] - List of masses of a cyclic peptide\n",
    "        Returns:\n",
    "            out: list[int] - Sorted list of the cyclospectrum of the input.\n",
    "        \"\"\"\n",
    "        seqCycle = seq*2\n",
    "        out = [0, sum(seq)]\n",
    "\n",
    "        for i in range(len(seq)):\n",
    "            for j in range(1, len(seq)):\n",
    "                subseq = seqCycle[i:i+j]\n",
    "                out.append(sum(subseq))\n",
    "        return sorted(out)\n",
    "\n",
    "    def linearSpectrum(self, peptide):\n",
    "        \"\"\"\n",
    "        Generates the linear spectrum given a list of masses of a linear peptide.\n",
    "        Parameters:\n",
    "            peptide: list[int] - List of masses of a linear peptide\n",
    "        Returns:\n",
    "            spec: list[int] - Sorted list of the linear spectrum of the input.\n",
    "        \"\"\"\n",
    "        spec = [0]\n",
    "        for x in range(1,len(peptide)):\n",
    "            for i in range(len(peptide)):\n",
    "                if i+x <= len(peptide):\n",
    "                    spec.append(sum(peptide[i:i+x]))\n",
    "        spec.append(sum(peptide))\n",
    "        \n",
    "        return sorted(spec)\n",
    "     \n",
    "    def isConsistent(self, peptide):\n",
    "        \"\"\"\n",
    "        Determines if candidate linear peptide is consistent with input spectrum\n",
    "\n",
    "        Parameters:\n",
    "            peptide: list[int] - List of masses of a linear peptide\n",
    "        Returns\n",
    "            bool: - A peptide is consistent if all of its masses are in the spectra being compared to.\n",
    "        \"\"\"\n",
    "        linSpec = self.linearSpectrum(peptide)\n",
    "        for p in linSpec:\n",
    "            if p not in self.spectrum:\n",
    "                return False\n",
    "        return True\n",
    "    \n",
    "    def expand(self, peps):\n",
    "        \"\"\"\n",
    "        Expands the peptides in input by 1 amino acid mass\n",
    "\n",
    "        Parameters:\n",
    "            peps: list[list[int]] - The current list of candidate peptides\n",
    "\n",
    "        Returns:\n",
    "            newPeps: list[list[int]] - All peptides in peps, but with 1 amino acid added.\n",
    "        \"\"\"\n",
    "        newPeps = []\n",
    "        for item in peps:\n",
    "            for mass in set(self.massTable.values()):\n",
    "                newPeps.append(item + [mass])\n",
    "        return newPeps\n",
    "    \n",
    "    def CylcopeptideSequencing(self):\n",
    "        \"\"\"\n",
    "        Determines every amino acid string Peptide such that Cyclospectrum(Peptide) = Spectrum (if such a string exists).\n",
    "        \"\"\"\n",
    "        result = set()\n",
    "        peptides = [[]]\n",
    "        \n",
    "        # only search while we have peptides to test\n",
    "        while peptides:\n",
    "            peptides = self.expand(peptides)  # branching step, add new peptides \n",
    "            for peptide in peptides: \n",
    "                # check if mass of current peptide is equal to input \n",
    "                if sum(peptide) == self.spectrum[-1]: \n",
    "                    # check if the cyclospectrums match\n",
    "                    if self.cylcoSpectrum(peptide) == self.spectrum:\n",
    "                        # we have a match so add it \n",
    "                        result.add(tuple(peptide))\n",
    "                    # regardless if matched or not, we get rid of the peptide from our search\n",
    "                    peptides = [p for p in peptides if p != peptide]\n",
    "                # check for consistency, if not then get rid of peptide.\n",
    "                elif not self.isConsistent(peptide):\n",
    "                    peptides = [p for p in peptides if p != peptide]\n",
    "             \n",
    "        return result\n",
    "  \n",
    "    \n",
    "def main(inFile = None):\n",
    "    '''\n",
    "    Do the main thing\n",
    "    '''\n",
    "    \n",
    "    with open(inFile, \"r\") as f:\n",
    "        line = f.readline().strip()\n",
    "    \n",
    "    peptide = Peptides(line)\n",
    "    masses = peptide.CylcopeptideSequencing()\n",
    "    output = [\"-\".join(map(str,p)) for p in masses]\n",
    "    \n",
    "    \n",
    "    with open(\"cmirchandani_rosalind18_out.txt\", \"w\") as out:\n",
    "        print(*output, sep=\" \")\n",
    "        print(*output, sep=\" \", file=out)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile = 'rosalind_ba4e.txt') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "paperback-warning",
   "metadata": {},
   "source": [
    "## Inspection Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "959ce21b",
   "metadata": {},
   "source": [
    "Inspection group: Gabe P., Jodie J.,\n",
    "\n",
    "- Added print to stdout as well as file\n",
    "- Added more inline comments to explain code \n",
    "- Used map() to convert input to ints, and to make output str"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13 (main, May 24 2022, 21:13:51) \n[Clang 13.1.6 (clang-1316.0.21.2)]"
  },
  "vscode": {
   "interpreter": {
    "hash": "b0fa6594d8f4cbf19f97940f81e996739fb7646882a419484c72d19e05852a7e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
